packages/astro/src/core/util.ts

```diff
@@ -120,11 +120,23 @@ function isInjectedRoute(file: URL, settings: AstroSettings) {
 }
 
 function isPublicRoute(file: URL, config: AstroConfig): boolean {
-	const pagesDir = resolvePages(config);
-	const parts = file.toString().replace(pagesDir.toString(), '').split('/').slice(1);
+	const rootDir = config.root.toString();
+	const pagesDir = resolvePages(config).toString();
+	const fileDir = file.toString();
+
+	// Normalize the file directory path by removing the pagesDir prefix if it exists,
+	// otherwise remove the rootDir prefix.
```

Comment:
The comment should also explain **why** this is necessary. The code itself isn't too complicated to understand, so a comment that *explains* seems superfluous. However, it's worth adding an explain of the motive, because if you left a comment here, it means there's a special case to handle :)

Reply:
> It might happen that a page is located in `node_modules`, and the Astro project itself is in a path containing the `_` character. In such cases, we should use a relative path to bypass the segment containing `_`.

WDYT?